src/commands/run.rs

```diff
@@ -494,36 +494,116 @@ impl RunCommand {
             }
             #[cfg(feature = "component-model")]
             CliLinker::Component(linker) => {
+                let component = main_target.unwrap_component();
                 if self.invoke.is_some() {
-                    bail!("using `--invoke` with components is not supported");
+                    self.invoke_component(&mut *store, component, linker).await
+                } else {
+                    let command = wasmtime_wasi::bindings::Command::instantiate_async(
+                        &mut *store,
+                        component,
+                        linker,
+                    )
+                    .await?;
+
+                    let result = command
+                        .wasi_cli_run()
+                        .call_run(&mut *store)
+                        .await
+                        .context("failed to invoke `run` function")
+                        .map_err(|e| self.handle_core_dump(&mut *store, e));
+
+                    // Translate the `Result<(),()>` produced by wasm into a feigned
+                    // explicit exit here with status 1 if `Err(())` is returned.
+                    result.and_then(|wasm_result| match wasm_result {
+                        Ok(()) => Ok(()),
+                        Err(()) => Err(wasmtime_wasi::I32Exit(1).into()),
+                    })
                 }
+            }
+        };
+        finish_epoch_handler(store);
 
-                let component = main_target.unwrap_component();
+        result
+    }
+
+    #[cfg(feature = "component-model")]
+    async fn invoke_component(
+        &self,
+        store: &mut Store<Host>,
+        component: &wasmtime::component::Component,
+        linker: &mut wasmtime::component::Linker<Host>,
+    ) -> Result<()> {
+        use wasmtime::component::{
+            types::ComponentItem,
+            wasm_wave::{
+                untyped::UntypedFuncCall,
+                wasm::{DisplayFuncResults, WasmFunc},
+            },
+            Val,
+        };
+
+        let invoke: &String = self.invoke.as_ref().unwrap();
+
+        // Check if input is wrapped in double quotes
+        let lacks_quotes = !invoke.starts_with('"') && !invoke.ends_with('"');
```

Comment:
I would like to revisit this logic (and will take some help if possible).
I discussed why Rust is not able to actually check for outer quotes in [this comment ](https://github.com/bytecodealliance/wasmtime/pull/10511#issuecomment-2774552109).
But now that I realize we can use single quotes to surround the whole function there might be another way to attempt to detect the absence of surrounding quotes (that are passed, to Rust, from the CLI). I will try and get back to this in a few hours and do some local testing.

Reply:
Updates are ready at
https://github.com/bytecodealliance/wasmtime/pull/10533